downsample_rs/src/m4/generic.rs

```diff
@@ -1,27 +1,29 @@
-// use ndarray::parallel::prelude::*;
 use ndarray::Zip;
 use ndarray::{s, Array1, ArrayView1};
 
+use rayon::iter::IndexedParallelIterator;
+use rayon::prelude::*;
+use std::sync::{Arc, Mutex};
+
+// --------------------- WITHOUT X
+
 #[inline(always)]
 pub(crate) fn m4_generic<T: Copy + PartialOrd>(
     arr: ArrayView1<T>,
     n_out: usize,
     f_argminmax: fn(ArrayView1<T>) -> (usize, usize),
 ) -> Array1<usize> {
-    // Non-parallel implementation
+    // Assumes n_out is a multiple of 4
     if n_out >= arr.len() {
         return Array1::from((0..arr.len()).collect::<Vec<usize>>());
     }
 
-    let block_size = (arr.len() - 1) as f64 / (n_out - 1) as f64 * 4.0;
+    let block_size = arr.len() as f64 / (n_out as f64) * 4.0;
     let block_size = block_size.floor() as usize;
 
-    let n_out_actual = (arr.len() / block_size) * 4 + 1;
+    let mut sampled_indices: Array1<usize> = Array1::<usize>::default(n_out);
 
-    let mut sampled_indices: Array1<usize> = Array1::<usize>::default(n_out_actual);
-
-    arr.slice(s![..block_size * (n_out_actual - 1) / 4])
-        .exact_chunks(block_size)
+    arr.exact_chunks(block_size)
```

Comment:
Guess this can result in index out of bounds issue;
- either slice the array up to `block_size * n_out / 4`
- or `.take(n_out / 4)`